### PR TITLE
[MAPSME-4213] [ios] Added banners support to search.

### PR DESCRIPTION
**Pull request by igrechuhin**

### CODE REVIEW

File: iphone/Maps/Core/Search/MWMSearch.mm [hunk 1]
```diff
@@ -270,6 +295,40 @@ + (void)clearFilter
   MWMSearch * manager = [MWMSearch manager];
   manager.filter = nil;
   [manager update];
+  [manager onSearchCompleted];
+}
+
+- (void)updateItemsIndex
+{
+  auto resultsCount = self->m_everywhereResults.GetCount();
```
Comment (syershov) [inline iphone/Maps/Core/Search/MWMSearch.mm]: auto const

File: iphone/Maps/Core/Search/MWMSearch.mm [hunk 2]
```diff
@@ -270,6 +295,40 @@ + (void)clearFilter
...
+  auto itemsIndex = [[MWMSearchIndex alloc] initWithSuggestionsCount:self.suggestionsCount
```
Comment (syershov) [inline iphone/Maps/Core/Search/MWMSearch.mm]: auto const

File: iphone/Maps/Core/Search/MWMSearch.mm [hunk 3]
```diff
@@ -208,30 +217,48 @@ + (void)searchQuery:(NSString *)query forInputLocale:(NSString *)inputLocale
...
 }
 
++ (id<MWMBanner>)adWithContainerIndex:(NSUInteger)index
+{
+  return [[MWMSearch manager].banners bannerAtIndex:index];
+}
+
++ (MWMSearchItemType)resultTypeWithIndex:(NSUInteger)index
+{
+  auto itemsIndex = [MWMSearch manager].itemsIndex;
```
Comment (VladiMihaylenko) [inline iphone/Maps/Core/Search/MWMSearch.mm]: const

File: iphone/Maps/Core/Search/MWMSearch.mm [hunk 4]
```diff
@@ -208,30 +217,48 @@ + (void)searchQuery:(NSString *)query forInputLocale:(NSString *)inputLocale
...
+  return [itemsIndex resultTypeWithIndex:index];
+}
+
++ (NSUInteger)containerIndexWithIndex:(NSUInteger)index
+{
+  auto itemsIndex = [MWMSearch manager].itemsIndex;
```
Comment (VladiMihaylenko) [inline iphone/Maps/Core/Search/MWMSearch.mm]: const

File: iphone/Maps/UI/Search/TableView/MWMSearchTableViewController.mm [hunk 5]
```diff
@@ -86,41 +74,66 @@ - (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger
...
+  {
+  case MWMSearchItemTypeRegular:
   {
-    auto tCell = static_cast<MWMSearchSuggestionCell *>(cell);
-    [tCell config:[self searchResultForIndexPath:indexPath]];
-    tCell.isLastCell = indexPath.row == [MWMSearch suggestionsCount] - 1;
+    auto cell = static_cast<MWMSearchSuggestionCell *>([tableView
+        dequeueReusableCellWithCellClass:[MWMSearchCommonCell class]
+                               indexPath:indexPath]);
+    auto result = [MWMSearch resultWithContainerIndex:containerIndex];
```
Comment (VladiMihaylenko) [inline iphone/Maps/UI/Search/TableView/MWMSearchTableViewController.mm]: auto const &

File: iphone/Maps/UI/Search/TableView/MWMSearchTableViewController.mm [hunk 6]
```diff
@@ -86,41 +74,66 @@ - (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger
...
+    auto ad = [MWMSearch adWithContainerIndex:containerIndex];
+    [cell configWithAd:ad containerType:MWMAdBannerContainerTypeSearch];
+    return cell;
+  }
+  case MWMSearchItemTypeSuggestion:
+  {
+    auto cell = static_cast<MWMSearchSuggestionCell *>([tableView
+        dequeueReusableCellWithCellClass:[MWMSearchSuggestionCell class]
+                               indexPath:indexPath]);
+    auto suggestion = [MWMSearch resultWithContainerIndex:containerIndex];
```
Comment (VladiMihaylenko) [inline iphone/Maps/UI/Search/TableView/MWMSearchTableViewController.mm]: auto const &

File: iphone/Maps/UI/Search/TableView/MWMSearchTableViewController.mm [hunk 7]
```diff
@@ -86,41 +74,66 @@ - (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger
...
-    [Statistics logEvent:kStatEventName(kStatSearch, kStatSelectResult)
-          withParameters:@{kStatValue : suggestionString, kStatScreen : kStatSearch}];
-    [delegate searchText:suggestionString forInputLocale:nil];
-  }
-  else if (cls == [MWMSearchCommonCell class])
+  case MWMSearchItemTypeRegular:
   {
     MWMSearchTextField * textField = delegate.searchTextField;
     [MWMSearch saveQuery:textField.text forInputLocale:textField.textInputMode.primaryLanguage];
+    auto result = [MWMSearch resultWithContainerIndex:containerIndex];
```
Comment (VladiMihaylenko) [inline iphone/Maps/UI/Search/TableView/MWMSearchTableViewController.mm]: auto const &

File: iphone/Maps/UI/Search/TableView/MWMSearchTableViewController.mm [hunk 8]
```diff
@@ -86,41 +74,66 @@ - (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger
...
     [delegate processSearchWithResult:result];
+    break;
+  }
+  case MWMSearchItemTypeMopub: break;
+  case MWMSearchItemTypeSuggestion:
+  {
+    auto suggestion = [MWMSearch resultWithContainerIndex:containerIndex];
```
Comment (VladiMihaylenko) [inline iphone/Maps/UI/Search/TableView/MWMSearchTableViewController.mm]: auto const &

File: iphone/Maps/Core/Search/SearchIndex.swift [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  private var items: [Item] = []
+
+  var count: Int {
+    return items.count
+  }
+
+  init(suggestionsCount: Int, resultsCount: Int) {
+    for index in 0..<resultsCount {
+      let type: MWMSearchItemType = index < suggestionsCount ? .suggestion : .regular
+      positionItems.append(PositionItem(item: Item(type: type, containerIndex: index), position: index))
```
Comment (syershov) [inline iphone/Maps/Core/Search/SearchIndex.swift]:
```swift
let item Item(type: type, containerIndex: index)
positionItems.append(PositionItem(item: item, position: index))
```

File: iphone/Maps/Core/Search/SearchIndex.swift [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    }
+    super.init()
+  }
+
+  func addItem(type: MWMSearchItemType, prefferedPosition: Int, containerIndex: Int) {
+    assert(type != .suggestion && type != .regular)
+    positionItems.append(PositionItem(item: Item(type: type, containerIndex: containerIndex), position: prefferedPosition))
```
Comment (syershov) [inline iphone/Maps/Core/Search/SearchIndex.swift]: Here too.

**Pull request merged**